### PR TITLE
fix: ADT hover considering only type or const len not lifetimes

### LINKED ISSUE #16963: Panic on hovering over type name with lifetime argument
When hovering over the name of a type with a lifetime parameter, a panic occurs and [this error](https://github.com/rust-lang/rust-analyzer/files/14781790/stderr.txt) is printed to stderr. VS Code shows a popup with a "Request textDocument/hover failed." message.

**rust-analyzer version**: 0.4.1899-standalone (4b33850c3 2024-03-26)

**rustc version**: rustc 1.79.0-nightly (47ecded35 2024-03-26)

**code snippet to reproduce**:
Create a new empty project, then paste the following snippet:
```rust
fn main() {
    println!("Hello, world!");
}

struct Pedro<'a> {
    hola: &'a str
}
```
Now hover your mouse over the word `Pedro` and the panic will occur.

### PR DESCRIPTION
**Pull request by dfireBird**
I feel like test doesn't do much. Please suggest if I should improve it?

Fixes #16963

### CODE REVIEW

File: crates/hir/src/lib.rs [hunk 1]
```diff
@@ -1418,7 +1418,8 @@ impl Adt {
     }
 
     pub fn layout(self, db: &dyn HirDatabase) -> Result<Layout, LayoutError> {
-        if db.generic_params(self.into()).iter().count() != 0 {
+        let generic_params = &db.generic_params(self.into());
+        if generic_params.iter().count() != 0 || generic_params.iter_lt().count() != 0 {
```
Comment (lnicola) [inline crates/hir/src/lib.rs:1422]: Nit: `.iter().next().is_some()`

Comment (dfireBird) [reply crates/hir/src/lib.rs:1422]: Yeah, that could be better, since count consumes the entire iterator.

**Review by lnicola — APPROVED** [review]

**Comment by lnicola** [comment]
Thanks!

@bors r+

**Comment by lnicola** [comment]
The test does its job (crash without the fix), it's fine.

File: crates/hir/src/lib.rs [hunk 2]
```diff
@@ -1418,7 +1418,8 @@ impl Adt {
     }
 
     pub fn layout(self, db: &dyn HirDatabase) -> Result<Layout, LayoutError> {
-        if db.generic_params(self.into()).iter().count() != 0 {
+        let generic_params = &db.generic_params(self.into());
+        if generic_params.iter().next().is_some() || generic_params.iter_lt().next().is_some() {
```
Comment (Veykril) [inline crates/hir/src/lib.rs:1422]: Having a `GenericParam::is_empty` method would probably be useful here (or given there is more than just params in it, a `GenericParam::has_params`)

Comment (dfireBird) [reply crates/hir/src/lib.rs:1422]: Should I open a PR with the proposed change of this and the proposed change in another PR?

Feel like a opening a lot of small PRs 😅.

Comment (Veykril) [reply crates/hir/src/lib.rs:1422]: Putting both into one is fine

**Pull request merged**